src/java.base/share/classes/sun/security/jca/JCAUtil.java

```diff
@@ -69,4 +71,25 @@ public static SecureRandom getSecureRandom() {
         return CachedSecureRandomHolder.instance;
     }
 
+    // called by sun.security.jca.Providers class when provider list is changed
+    static void clearDefSecureRandom() {
+        def = null;
+    }
+
+    /**
+     * Get the default SecureRandom instance. This method is the
+     * optimized version of "new SecureRandom()" which re-uses the default
+     * SecureRandom impl if the provider table is the same.
+     */
+    public static SecureRandom getDefSecureRandom() {
+        if (def == null) {
+            synchronized (JCAUtil.class) {
+                if (def == null) {
+                    def = new SecureRandom();
+                }
+            }
+        }
+        return def;
```

Comment:
Do we need to worry about `clearDefSecureRandom()` being called between lines 88 and 92?

Thread 1: `getDefSecureRandom()#85` (`def` is `null`)
Thread 1: `getDefSecureRandom()#86`
Thread 1: `getDefSecureRandom()#87` (`def` is still `null`)
Thread 1: `getDefSecureRandom()#88` (`def` is **not** `null`)
Thread 2: `clearDefSecureRandom()` (`def` is `null` again)
Thread 1: `getDefSecureRandom()#92` (return `def` which is incorrectly `null`)


```suggestion
        SecureRandom result = def;
        if (result == null) {
            synchronized (JCAUtil.class) {
                result = def;
                if (result == null) {
                    def = result = new SecureRandom()
                }
            }
        }
        return result;
```

Reply:
Sure, good point. I will update. Thanks for catching this.